modal/_container_entrypoint.py

```diff
@@ -441,20 +553,110 @@ def import_function(
     else:
         obj = None
 
+    if function:
+        code_deps = function.deps(only_explicit_mounts=True)
+
     return ImportedFunction(
         obj,
         user_defined_callable,
-        function_def.webhook_config,
         active_app,
-        is_async,
-        is_generator,
-        data_format,
-        input_concurrency,
-        function_def.is_auto_snapshot,
-        function,
+        code_deps,
+    )
+
+
+def import_class_service(
+    function_def: api_pb2.Function,
+    ser_cls,
+    ser_fun,
+    ser_params: Optional[bytes],
+    client: Client,
+) -> Service:
+    """
+    This imports a full class to be able to execute any @method or webhook decorated methods.
+
+    See import_function.
+    """
+    module: Optional[ModuleType] = None
+    cls: Optional[Type] = None
+    active_app: Optional[_App] = None
+    code_deps: Optional[List["modal.object._Object"]] = None
+
+    if function_def.definition_type == api_pb2.Function.DEFINITION_TYPE_SERIALIZED:
+        assert ser_cls is not None
+        cls = ser_cls
+        # This is a serialized function we already fetched from the server
+        method_partials = _find_partial_methods_for_cls(cls, _PartialFunctionFlags.all())
+    else:
+        # Load the module dynamically
+        module = importlib.import_module(function_def.module_name)
+        qual_name: str = function_def.function_name
+
+        if not is_global_object(qual_name):
+            raise LocalFunctionError("Attempted to load a class defined in a function scope")
+
+        parts = qual_name.split(".")
+        if not (
+            len(parts) == 2 and parts[1] == "*"
+        ):  # the "function name" of a class service "function placeholder" is expected to be "ClassName.*"
+            raise InvalidError(f"Invalid 'class function' identifier {qual_name}")
+
+        assert not function_def.use_method_name  # new "placeholder methods" should not be invoked directly!
+        cls_name = parts[0]
+        cls = getattr(module, cls_name)
+        if isinstance(cls, Cls):
+            # The cls decorator is in global scope
+            _cls: "modal.cls._Cls" = synchronizer._translate_in(cls)
+            method_partials = _cls._method_partials  # implementation to follow in cls refactor PR
+            active_app = _cls._app
+            # TODO: set code deps using the decorator options of the cls - maybe using _cls.deps?
+        else:
+            # This is a raw class - find all methods
+            method_partials = _find_partial_methods_for_cls(cls, _PartialFunctionFlags.all())
+
+    # Instantiate the class if it's defined
+    assert cls  # must be a class
+    if ser_params:
+        _client: _Client = synchronizer._translate_in(client)
+        args, kwargs = deserialize(ser_params, _client)
+    else:
+        args, kwargs = (), {}
+    obj = cls(*args, **kwargs)
+    if isinstance(cls, Cls):
+        obj = obj.get_obj()
```

Comment:
What's going on here?

Reply:
Yeah this is quite hard to grok, and we should try to clarify the code a bit IMO (this is mostly as is from the pre-patch code).

If you have a `modal.Cls` instance (`cls` in this case, so `Cls` can basically be considered a metaclass...) and `__call__` it (which to a user looks like an instantiation of the class), it returns an `Obj` instance, which is not an actual instance of the user's defined class (the one decorated) but a lazily evaluated placeholder for the args and kwargs that should be used to create the actual instance of the user's class.

`obj.get_obj()` in this case takes the `Obj` and returns an instance of the user's actual class. The reason that it's if-gated is due to the imported `cls` not always necessarily being a `modal.Cls` instance. In case the class isn't decorated in global scope it could **be** the actual user class, in which case we wouldn't have to do the whole `.get_obj()` thing to get to the underlying instance.

Might actually take a stab at putting better names on these entities/methods now that I have it fresh in my head, as it's definitely confusing...